src/services/battleye/battleye_service.cpp

```diff
@@ -0,0 +1,247 @@
+
+		// send to player
+		if (auto plyr = g_player_service->get_by_host_token(token); plyr && plyr->get_session_player())
+		{
+			pkt.send(plyr->get_session_player()->m_msg_id, true);
+		}
+	}
+
+	void battleye_service::kick_player(std::uint64_t token, const char* reason)
+	{
+		if (auto plyr = g_player_service->get_by_host_token(token))
+		{
+			LOGF(WARNING, "BattlEye wants us to kick {} for {}", plyr->get_name(), reason);
+			if (!plyr->tampered_with_be)
+				session::add_infraction(plyr, Infraction::TRIGGERED_ANTICHEAT);
+		}
+	}
+
+	void battleye_service::script_func()
+	{
+		bool was_host = false;
+		while (is_running())
+		{
+			if (g_player_service->get_self()->is_valid())
+			{
+				bool is_host = g_player_service->get_self()->is_host();
+
+				if (is_host != was_host)
+				{
+					if (is_host)
+					{
+						for (auto& [_, player] : g_player_service->players())
+						{
+							add_player(player->get_net_game_player()->get_host_token(), player->get_rockstar_id(), player->get_name());
+						}
+					}
+					was_host = is_host;
+				}
+			}
+
+			big::script::get_current()->yield();
+		}
+	}
+
+	void battleye_service::thread_func()
+	{
+		while (is_running() && g_running)
+		{
+			// we need to run this every frame (lockstep with game or otherwise)
+			{
+				std::lock_guard lock(m_mutex);
+				if (is_running() && !m_battleye_api.m_run())
+					LOG(WARNING) << "BE::Run failed";
+			}
+
+			std::this_thread::yield();
+		}
+	}
+
+	battleye_service::battleye_service()
+	{
+	}
+
+	battleye_service::~battleye_service()
+	{
+		stop();
+	}
+
+	void battleye_service::start()
+	{
+		std::lock_guard lock(m_mutex);
+
+		if (is_running())
+			return;
+
+		auto handle = LoadLibraryA("BattlEye\\BEServer_x64.dll");
+
+		if (!handle)
+		{
+			LOG(WARNING) << "Failed to load BattlEye";
+			return;
+		}
+
+		m_battleye_user_data.m_game_name = "paradise";
+		m_battleye_user_data.m_log_func  = [](const char* message, int level) {
+            LOGF(INFO, "BattlEye: {}", message);
+		};
+		m_battleye_user_data.m_kick_player = [](std::uint64_t player, const char* reason) {
+			g_battleye_service.kick_player(player, reason);
+		};
+		m_battleye_user_data.m_send_message = [](std::uint64_t player, const void* pkt_data, int size) {
+			g_battleye_service.send_message(player, const_cast<void*>(pkt_data), size);
+		};
+		if (reinterpret_cast<init_t>(GetProcAddress(handle, "Init"))(1, &m_battleye_user_data, &m_battleye_api))
+		{
+			// background thread (not game thread)
+			g_thread_pool->push([this] { this->thread_func(); });
+
+			// background script (game thread)
+			g_script_mgr.add_script(std::make_unique<big::script>([this] { this->script_func(); }, "BE Background Script", false));
+		}
+		else
+		{
+			LOG(WARNING) << "Failed to initialize BattlEye";
+		}
+	}
+
+	void battleye_service::stop()
+	{
+		std::lock_guard lock(m_mutex);
+
+		if (!is_running())
+			return;
+
+		m_battleye_api.m_shutdown();
+		memset(&m_battleye_api, 0, sizeof(CApi));

```

Comment:
You are zeroing out the memory, why not just delete it or release it with a smart ptr?